backends/bmv2/common/parser.cpp

```diff
@@ -36,9 +72,10 @@ cstring ParserConverter::jsonAssignment(const IR::Type *type, bool inParser) {
         else
             return "assign_header_stack";
     }
-    if (inParser)
+    if (inParser && isFieldExpr(expr))
         // Unfortunately set can do some things that assign cannot,
         // e.g., handle lookahead on the RHS.
+        // One limitation of set is that its LHS has to be a field expr.
```

Comment:
my understanding is that at this stage, we are guaranteed that if there is a `lookahead` on the RHS, the LHS has to be a field expression. @mihaibudiu mentioned that there was a pass to extract all `lookahead`s and assign their value to a temporary metadata field.

Maybe the comment should be updated a bit:
> Unlike "assign", "set" is implemented more efficiently in bmv2, and can handle a lookahead expression as the RHS. However, "set" requires its LHS to be a field expression.
> The RemoveComplexExpressions pass guarantees that all lookahead<T>() expressions are lifted into tmp = lookahead<T>(). Therefore, we are guaranteed that anytime the RHS includes a lookahead, it will be a bare lookahead call and the LHS will be a simple field expression. These assignments will be handled by this if branch, using "set".

Reply:
Updated.